autoscale: re-constructed autoscaling module

romberli opened the pull request:
<!--
Thank you for working on PD! Please read PD's [CONTRIBUTING](https://github.com/tikv/pd/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed


If you want to open the **Challenge Program** pull request, please use the following template:
https://raw.githubusercontent.com/tikv/.github/master/.github/PULL_REQUEST_TEMPLATE/challenge-program.md
You can use it with query parameters: https://github.com/tikv/pd/compare/master...${you branch}?template=challenge-program.md
-->

### What problem does this PR solve?

<!-- Add the issue link with a summary if it exists. -->
Reconstructed the autoscaling module

### What is changed and how it works?

Now autoscaling has two different types: homogeneous scaling and heterogeneous scaling.

* If average cpu/storage usage is larger than MaxThreshold, homogeneous scaling will be performed.
* If average cpu usage is not larger than MaxThreshold, but cpu usages of some instances are larger than MaxThreshold,  heterogeneous scaling will be performed.
* Now strategy will include kubernetes node count
* Now MaxThreshold of storage auto rule means maximum storage usage
* Added MinThreshold to storage auto rule

### Check List

<!-- Remove the items that are not applicable. -->

Tests

<!-- At least one of them must be included. -->

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has configuration change
- Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tidb-ansible`](https://github.com/pingcap/tidb-ansible):
- Need to cherry-pick to the release branch

### Release note

<!-- A bugfix or a new feature needs a release note. If there is no need release note, just uncomment the below line. -->
```release-note
Please add a release note.

If you don't think this PR needs a release note then fill it with None.
```

rleungx commented:
Does this PR have a test result in the real cluster?

romberli commented:
> Does this PR have a test result in the real cluster?

@rleungx , yes, we've tested in our test environment for a while

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -14,26 +14,23 @@
 )
 
 const (
+	DefaultTimeout                 = 5 * time.Second
```
```suggestion
	defaultTimeout                 = 5 * time.Second
```

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,363 @@ var (
+		resourceMap map[string]uint64
+	)
+	// init prometheus client
+	address, err := getPrometheusAddress(rc)
```
Why not use `cfg.MetricStorage`?

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,363 @@ var (
+		return nil, err
+	}
+
+	log.Debug(fmt.Sprintf("prometheus address: %s", address))
```
How about using zap for all logs? You can refer to other logs in our repo.

rleungx commented on pkg/autoscaling/calculation_test.go:
```diff
@@ -17,360 +17,309 @@ import (
-	totalCPUUseTime, _ := getTotalCPUUseTime(querier, TiDB, instances, time.Now(), 0)
-	expected := mockResultValue * float64(len(instances))
-	c.Assert(math.Abs(expected-totalCPUUseTime) < 1e-6, IsTrue)
+func Test(t *testing.T) {
```
prefer to move it to the head of this file

rleungx commented on pkg/autoscaling/handler.go:
```diff
@@ -50,12 +52,26 @@ func (h *HTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	data, err = json.Marshal(plans)
+	if err != nil {
+		log.Error("marshal plans failed.", errs.ZapError(err))
```
need to return an error?

nolouch commented:
Because this feature is more than in the future, we first take it to the branch `auto-scaling-impRovement`.

nolouch commented:
ci failed @romberli

rleungx commented on pkg/autoscaling/types.go:
```diff
@@ -98,6 +166,10 @@ const (
 	CPUQuota
+	// StorageUsage is used storage usage in the duration
+	StorageUsage
+	// StorageQuota is cpu storage quota for each instance
```
cpu storage？

romberli replied:
done

rleungx commented on pkg/autoscaling/handler.go:
```diff
@@ -50,12 +52,30 @@ func (h *HTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	log.Debug("autoscaling: get http body completed", zap.String("strategy", string(data)))
```
```suggestion
	log.Debug("get http body completed", zap.String("strategy", string(data)))
```
I think `autoscaling:` is unnecessary

romberli replied:
done

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,399 @@ var (
-		case "tikv":
-			components[TiKV] = struct{}{}
+		case TiKV.String():
+			// get heterogeneous groups
```
We can abstract a function here.

romberli replied:
done

romberli replied:
done

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,399 @@ var (
+		return "", err
+	}
+	if len(resp.Kvs) == 0 {
+		return "", errors.New(fmt.Sprintf("length of the response values of the key %s is 0", prometheusAddressKey))
```
How about using `errors.Errorf` for all `errors.New(fmt.Sprintf())`?

romberli replied:
done

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,399 @@ var (
+	return getCPUPlans(querier, instances, strategy, resourceMap, TiDB)
+}
+
+func getCPUPlans(querier Querier, instances []instance, strategy *Strategy, resourceMap map[string]uint64, component ComponentType) ([]*Plan, error) {
```
How about directly using getTiDBPlans?

romberli replied:
did you mean change getCPUPlans() function name to getTiDBPlans ?
if yes, I'm afraid we cann't.because getCPUPlans() function not only used by getTiDBPlans() function but also used by getTiKVPlans() function, and calling getTiDBPlans() function in getTiKVPlans() function is meaningless, it doesn't make sense.

rleungx replied:
Oh, you are right.

rleungx commented on tests/prometheus/service.go:
```diff
@@ -0,0 +1,48 @@
+// Copyright 2020 TiKV Project Authors.
```
```suggestion
// Copyright 2021 TiKV Project Authors.
```

romberli replied:
done

rleungx commented on tests/prometheus/service.go:
```diff
@@ -0,0 +1,48 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
```
This license is short of a sentence, please use the latest one.

romberli replied:
short of what? I copied this from other file header, could you tell me a file path that I can copied from?

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -143,298 +450,386 @@ func filterTiKVInstances(informer core.StoreSetInformer) []instance {
 	switch component {
 	case TiKV:
-		return getScaledTiKVGroups(rc, healthyInstances)
+		resourceMap, err = getTiKVResourceMap(rc, healthyInstances)
```
How about `return getTiKVResourceMap(rc, healthyInstances)`?

romberli replied:
done

rleungx commented on pkg/autoscaling/types.go:
```diff
@@ -69,6 +91,34 @@ type Plan struct {
+	return fmt.Sprintf("{Component: %s, Count: %d, ResourceType: %s, Labels: %v}", p.Component, p.Count, p.ResourceType, p.Labels)
+}
+
+func (p *Plan) Clone() *Plan {
```
CloneWithoutLabel?

romberli replied:
done

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,399 @@ var (
-	maxThreshold, minThreshold := getCPUThresholdByComponent(strategy, component)
+	totalInstanceCount := uint64(len(instances))
+	totalCPUUsage := totalCPUUsedTime / totalCPUQuota
+	cpuUsageTarget := (cpuMaxThreshold + cpuMinThreshold) / 2
```
Is it reasonable?

romberli replied:
which one? sorry, I don't follow

rleungx replied:
`cpuUsageTarget` is calculated through `cpuMaxThreshold` and `cpuMinThreshold`

romberli replied:
it's reansonable.
cpuUsageTarget is the target cpu usage percents that we want tikv/tidb instances stays,
`(cpuMaxThreshold + cpuMinThreshold) / 2`, it will return the middle value between `cpuMaxThreshold` and `cpuMinThreshold`, this value is neither too big nor too small.

rleungx commented:
What is the difference between homogeneous and heterogeneous nodes?

rleungx commented on tests/autoscaling/autoscaling_test.go:
```diff
@@ -40,18 +48,29 @@ type apiTestSuite struct{}
 	c.Assert(err, IsNil)
-	cluster.WaitLeader()
+	leader := cluster.WaitLeader()
+	fmt.Printf("leader: %s\n", leader)
```
remove it?

romberli replied:
done

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,399 @@ var (
+	for resourceType, resourceCount := range resourceMap {
+		if resourceType != homogeneousTiKVResourceType && resourceType != homogeneousTiDBResourceType {
+			plans = append(plans, NewPlan(component, resourceCount-1, resourceType))
+			log.Debug("get heterogeneous scale in plans completed",
```
break here to avoid the duplicated code?

romberli replied:
done

rleungx commented on pkg/autoscaling/calculation.go:
```diff
@@ -52,89 +49,399 @@ var (
+
+	switch component {
+	case TiKV:
+		instances = getTiKVInstances(rc)
```
directly use return?

romberli replied:
done

romberli commented:
> ci failed @romberli

in the ci log, it shows "undefined: io.ReadAll"...etc,
is this related to the go environment of the CI?

rleungx commented:
> > ci failed @romberli
> 
> in the ci log, it shows "undefined: io.ReadAll"...etc,
> is this related to the go environment of the CI?

let me check

purelind commented:
/rebuild

romberli commented:
> What is the difference between homogeneous and heterogeneous nodes?
please refer to the [document](https://github.com/romberli/tidb-operator/blob/master/docs/design-proposals/2021-07-22-auto-scaling-with-pdapi.md)

rleungx commented:
Please sign off DCO.

romberli commented:
> Please sign off DCO.

I signed DCO for all my code.
but after fetching upstream, it asks me to sign again, I'm not familiar with the rule, but anyway, I followed the instructions:
git rebase HEAD~28 --signoff
git push --force-with-lease origin master

and now , it turns out that branch has conflicts that must be resolved,
should the auto-scaling-improvement branch fetch master branch too?

nolouch commented:
I think some wrongs in `rebase`, there are 10000+ diff lines:( 
maybe you need to extract your code into a new commit, base the `auto-scaling-improvement`  branch.

romberli commented:
> improvement

I think it's related to the fetching upstream, maybe merging master branch to auto-scaling-improvement branch solves the problem

romberli commented:
@rleungx @nolouch ,
because fetching upstream caused merging conflict, and there too many commits to reverse,
I think it's better to create a new pull request, I deleted my fork repository, forked tikv/pd repository again, and copied files that I modified before back to the auto-scaling-improvement branch.
and I've created a new pr #4180 , could we use that pr to review?

rleungx commented:
> @rleungx @nolouch , because fetching upstream caused merging conflict, and there too many commits to reverse, I think it's better to create a new pull request, I deleted my fork repository, forked tikv/pd repository again, and copied files that I modified before back to the auto-scaling-improvement branch. and I've created a new pr #4180 , could we use that pr to review?

Sure!

romberli commented:
as we use pr #4180 , I'll close this pr
